### ISSUE TITLE
The disk guard blocks the corpus-scale runs that streaming and chunking exist to enable

### ISSUE THREAD

**Issue by thodel** [thodel/training-atr-models#26]
> **Stand 23.09.2026.** Die vier Punkte unter `## Scope` sind umgesetzt (900d2b2, siehe den Kommentar unten); geprüft gegen `main` (`415c5e0`) in `src/atr_training/hf_source.py` und `src/atr_training/ketos_cmd.py`. Offen war nur noch der Nachtrag unter „Deliberately not in scope" — **der steht jetzt als eigene Issue #72**, mit den Invarianten, an denen ein erster Anlauf (#66) gescheitert ist: ein vorhandenes `train_<i>.arrow` ist kein Beweis, dass sein `ketos compile` mit 0 geendet hat, und `pages_train_<i>.lst` enthält Seiten, keine Zeilen. Diese Issue selbst hat für den Bot nichts mehr zu tun und trägt die Marke `dh-bot` nicht mehr; sie kann geschlossen werden, sobald #72 gelandet ist.

> **Moved from thodel/serving-atr-inference#85** on 17.09.2026. Since the split (serving-atr-inference#137, serving-atr-inference#139) the training code lives in this repository: paths under `src/atr_serving/training/` are `src/atr_training/` here. Bare issue numbers below were rewritten to `serving-atr-inference#N`. Original, with its full history: https://github.com/thodel/serving-atr-inference/issues/85
>
> **State on 17.09.2026** (checked against the code): The four scope items are done (900d2b2, 70cacf7, b480ead) and carried over to training-atr-models. hf_source.py:700-757 _oversize_error refuses on parquet size only with caching on, refuses unchunked streaming and names ATR_TRAIN_CACHE_DATASETS / ATR_TRAIN_CHUNK_PAGES, and refuses engines that do not chunk. ketos_cmd.py:48-67 compile_workers tapers the worker count and is used at kraken runner.py:72-89. Still open, per the owner comment: runner.py:188-253 _compile_chunked always starts at chunk 0 and never checks for an existing train_<i>.arrow. The trainer API (app.py:616-957) has no retry or requeue for a failed job, and settings.py:342 still defaults chunk_pages to 0. New gap I found: app.py:760 sets chunk_capable from the engine alone, so a multi-dataset kraken request with eval_projects and chunk_pages>0 passes the guard (hf_source.py:745-746). But runner_base.py:253-261 sends any request with more than one dataset to _prepare_multi, and that path never chunks (runner_base.py:597-625). The guard also compares each spec separately against min_free_disk_gb (hf_source.py:869), never the sum.

---

Filed from the post-mortem of `20260808T183111Z-kraken-medieval-full-v1`, which
spent 23 h in `prepare` and then died in `compile`:

```
prepare:  461,586 pages, 15,406,203 lines   (23 h, ~526 GB)
compile:  StageFailed — ketos exited -9     (after 1 h 51 m)
```

Exit -9 is SIGKILL. The leading hypothesis is the kernel OOM killer — a single
`ketos compile` over a 461 K-page manifest with `--workers 8` — but it is **not
confirmed**: `journalctl -k` implies `-b`, so a reboot since would hide it, and the
system journal needs privileges. The 111-minute climb before death fits a memory
leak, and the alternative (a service restart reaching the detached child) is ruled
out: the restarts were on 08-13, four days later.

**The OOM itself is probably already fixed.** serving-atr-inference#39 landed chunked
materialize → compile → discard on 08-10, and `_compile_chunked`
(`engines/kraken_train_svc/runner.py:104`) compiles ≤ `chunk_pages` at a time, so no
single `ketos` invocation sees the whole manifest. What this issue is about is that
**you cannot get to that path.**

## The defect

Three settings that should cooperate do not know about each other:

| setting | default | what it does |
|---|---|---|
| `cache_datasets` | **`False`** | stream parquet instead of downloading it |
| `chunk_pages` | **`0`** (off) | bound peak page-disk to one chunk |
| `min_free_disk_gb` | `50` | refuse a selection larger than this |

The size guard at `src/atr_serving/training/hf_source.py:528-552` sizes the whole
parquet selection and refuses if it exceeds `min_free_disk_gb` — **regardless of
`cache_datasets`**. But with streaming on, which is the default, those shards are
never all resident. The guard measures a quantity the configured pipeline never
materializes, and rejects on it.

The result is a closed loop: a corpus-scale selection is refused for a download that
will not happen, and the two settings that make it feasible are invisible. The
error names neither:

> the selection is ~1023.0 GB across 766 parquet shards, over the 50 GB the trainer
> keeps free. **Lower max_pages or free space.**

Lowering `max_pages` and freeing space are the two remedies that do not address it.
`ATR_TRAIN_CACHE_DATASETS=false` and `ATR_TRAIN_CHUNK_PAGES=5000` are the ones that
do, and neither is mentioned.

## And chunking is dormant

`chunk_pages: int = 0` (`settings.py:109`) means `_should_chunk`
(`runner_base.py:204`) returns `False` unless someone sets the env var. So a
corpus-scale job submitted today — if it got past the guard — would take the **same
unchunked path that died on 08-09**. The serving-atr-inference#39 machinery is complete, tested, and off.

That is also why `kraken-medieval-shard00-std` is being run by hand: sharding the
corpus manually is currently easier than switching on the feature that does it.

## Scope

- [ ] The size guard applies the parquet check only when `cache_datasets` is on.
- [ ] When streaming *unchunked* at scale, refuse with the remedy that works — page
      disk is still unbounded there, which is the 526 GB that accumulated over 23 h.
- [ ] The message names `ATR_TRAIN_CACHE_DATASETS` and `ATR_TRAIN_CHUNK_PAGES`.
- [ ] `--workers` scales down as the manifest grows, rather than a fixed 8.

## Deliberately not in scope — follow-up

**Chunked compile has no resume.** `_compile_chunked` writes `train_<i>.arrow` per
chunk and they persist, but nothing reads them back: a failure in chunk 40 restarts
from chunk 0 on a fresh job id, re-streaming and re-compiling everything. Chunking
converts one enormous loss into several smaller ones without making any of them
recoverable. `train_<i>.arrow` already existing is the natural resume marker.

Whether `chunk_pages` should default to non-zero is also left open: it changes the
path every job takes, and the 238-page test case does not need it.

---

## Comments in serving-atr-inference

**thodel, 2026-08-14**

Landed in 900d2b2 (on `main`). Four of the five scope items:

- [x] The parquet check applies only when `cache_datasets` is on.
- [x] Streaming *unchunked* at scale is still refused — page-disk is genuinely
      unbounded there — but with the remedy that works.
- [x] The message names `ATR_TRAIN_CACHE_DATASETS` and `ATR_TRAIN_CHUNK_PAGES`.
- [x] `compile --workers` tapers inverse-linearly above 20 K pages, never below 1,
      unchanged below it. `compile_workers(8, 461_586) == 1`.

A fifth case turned up while writing it: `chunk_pages` set **without**
`eval_projects` is inert, because `_should_chunk` needs them — a validation set
cannot come from splitting a stream that is discarded as it is read. That now
refuses with the actual reason instead of silently materializing everything.

693 tests pass, verified against the committed tree in a clean worktree rather than
the working copy.

**Still open, and the more valuable half:** chunked compile has no resume. Chunking
turns one enormous loss into several smaller ones without making any recoverable —
a failure in chunk 40 restarts from chunk 0 under a fresh job id, re-streaming and
re-compiling everything, even though every `train_<i>.arrow` is sitting on disk.
Whether `chunk_pages` should default to non-zero is also still open; it changes the
path every job takes.

**Comment by thodel** [comment]
## Erledigt — der Umfang ist umgesetzt, der Rest steht als #72

Der Kopf dieses Issues hält es seit dem 23.09.2026 selbst fest: die vier Punkte unter `## Scope` sind mit `900d2b2` umgesetzt und gegen `main` in `src/atr_training/hf_source.py` und `src/atr_training/ketos_cmd.py` geprüft. Offen blieb nur der Nachtrag unter „Deliberately not in scope" — das Resume für den gestückelten Compile —, und der wird als eigenes Issue #72 geführt, samt den Invarianten, an denen der erste Anlauf (#66) gescheitert ist.

Hier ist damit nichts mehr zu tun. Geschlossen; #72 trägt den Rest.
